### ISSUE TITLE
[RFP] Can large models do a simple task (i.e arithmetic) perfectly?

### ISSUE THREAD

**Issue by leogao2** [EleutherAI/project-menu#22]
## Background

If really overpowered models can't learn something basic like arithmetic well (assuming all-arithmetic training set, sane tokenization w/out BPEs, etc), that implies something about what LMs can and can't learn. If present day models can't learn a simple task with the huge advantage of massive overparameterization and a purely-task dataset and favorable tokenization, then we should be skeptical that future LMs will be superhuman at language tasks. Somewhat tangentially related in spirit to #21.

## What to plot?

For each of `{+, -, *, /} x [1, 10] digits`, train like a 6B+ model with character level tokenization on just math problems. Plot accuracy+perplexity for each. Also would be interesting to look at our of distribution generalization (i.e train on 5 digits, test on 6).

Maybe also repeat for a bunch of smaller models too, so we have a trendline to look at.

Then maybe we can also pick the model apart to see why it's failing at certain tasks. I bet for the multiplication tasks in particular the number of layers will be very important since there's a certain number of sequential steps tat need to happen.

**Comment by uSaiPrashanth** [comment]
[Mathematics Dataset](https://github.com/deepmind/mathematics_dataset)

**Comment by cfoster0** [comment]
Prior art, for addition and subtraction: 
Investigating the Limitations of Transformers with Simple Arithmetic Tasks
https://arxiv.org/abs/2102.13019

**Comment by tnwei** [comment]
> Prior art, for addition and subtraction: Investigating the Limitations of Transformers with Simple Arithmetic Tasks https://arxiv.org/abs/2102.13019

The paper's authors concluded from their experiments that models couldn't learn addition rules independent of the length of numbers seen in training, is that observation concrete enough to answer the RFP?

**Comment by iRonJ** [comment]
We know LLMs can learn to write code, I suspect they would do extremely well on arithmetic if you just do prompt engineering such as "//javascript to solve 3+5-3" or something along those lines

In the vein of this problem though, thinking about how DeepMind used database lookups to shrink their LM and boost performance (https://deepmind.com/blog/article/language-modelling-at-scale), what if you had 2 smaller language models. One trained to automatically identify math problems, and the other trained to internally generate the prompt to solve those math problems, then executing the code to solve the math problems outside of the neural net, but passed back to it, then a final language model reads the original question, the predicted math equation, and the generated answer.

![Untitled-2022-03-19-1834](https://user-images.githubusercontent.com/6108031/159148490-31e9104c-aa8e-4ac2-9702-4aed929b5a99.png)

Each submodel can be trained independently of the others. The model at step 4 would most resemble existing language models, but would not be attempting to generate an answer like current LLMs, but instead relying on the submodels to generate the answer. As research advances, the symbolic programming interpreter could be replaced with perhaps a fully neural module, or new type of neural network that specializes in program generation.

**Comment by iRonJ** [comment]
> Each submodel can be trained independently of the others. The model at step 4 would most resemble existing language models, but would not be attempting to generate an answer like current LLMs, but instead relying on the submodels to generate the answer. As research advances, the symbolic programming interpreter could be replaced with perhaps a fully neural module, or new type of neural network that specializes in program generation.

The other idea behind the sub models is that something like clip can be plugged in and trained to similar text encodings, and the same math submodels should transfer to VQA tasks.

**Comment by iRonJ** [comment]
This approach would also allow for multi-step computations

**Comment by iRonJ** [comment]
https://arxiv.org/abs/2203.13224

An interesting paper here that I think could inform an approach to this issue

**Comment by iRonJ** [comment]
looks like Google took a whack at arithmetic by adding a calculator

http://ai.googleblog.com/2022/04/pathways-language-model-palm-scaling-to.html

>For example, with 8-shot prompting, PaLM solves 58% of the problems in [GSM8K](https://github.com/openai/grade-school-math), a benchmark of thousands of challenging grade school level math questions, outperforming the [prior top score](https://arxiv.org/abs/2110.14168) of 55% achieved by fine-tuning the GPT-3 175B model with a training set of 7500 problems and combining it with an external calculator and verifier.

**Comment by iRonJ** [comment]
Further work here, related to robotics, and having models talk to each other, with a very similar circular diagram as above:
https://socraticmodels.github.io/


![](https://socraticmodels.github.io/images/SM-discussion.png)


link to paper https://arxiv.org/abs/2204.00598

**Comment by iRonJ** [comment]
some other work in this area

https://www.ai21.com/blog/jurassic-x-crossing-the-neuro-symbolic-chasm-with-the-mrkl-system

**Comment by iRonJ** [comment]
Here’s a user of gpt neo that uses an extensive prompt to get reliable arithmetic

https://news.ycombinator.com/item?id=30309302

**Comment by StellaAthena** [comment]
> Here’s a user of gpt neo that uses an extensive prompt to get reliable arithmetic
> 
> https://news.ycombinator.com/item?id=30309302

This is GPT-3, presumably the 175B model, not GPT-Neo

**Comment by manuelsh** [comment]
> > Prior art, for addition and subtraction: Investigating the Limitations of Transformers with Simple Arithmetic Tasks https://arxiv.org/abs/2102.13019
> 
> The paper's authors concluded from their experiments that models couldn't learn addition rules independent of the length of numbers seen in training, is that observation concrete enough to answer the RFP?

I think it is. What additional observations are needed?

**Comment by toontran** [comment]
> > > Prior art, for addition and subtraction: Investigating the Limitations of Transformers with Simple Arithmetic Tasks https://arxiv.org/abs/2102.13019
> > 
> > The paper's authors concluded from their experiments that models couldn't learn addition rules independent of the length of numbers seen in training, is that observation concrete enough to answer the RFP?

> I think it is. What additional observations are needed?

This paper reminds me of the following one, where the author claims that BERT models cannot deductively reason, i.e. use rules, due to "statistical features".
https://arxiv.org/abs/2205.11502

But then we saw how GPT-3 can do arithmetic reasonably well if steps are explicitly written out
> > Here’s a user of gpt neo that uses an extensive prompt to get reliable arithmetic
> > 
> > https://news.ycombinator.com/item?id=30309302

> This is GPT-3, presumably the 175B model, not GPT-Neo

Is there a way to reconcile both these facts? I know this is not an apple-to-apple comparison, but it seems that models would have to elaborate steps sequentially in order to reason in general, instead of shoving everything into input and expecting correct output all at once.

**Comment by EDmitry** [comment]
What's curious is that perhaps LLMs can interact with external systems just like humans can:
<img width="722" alt="Screenshot 2022-12-06 at 10 03 01 AM" src="https://user-images.githubusercontent.com/141007/205987732-5fe67585-ad4c-43b7-909d-7baf308db849.png">
But that starts from making it understand its own limitations, which is what I've done here with this prompt.
